bin/pytorch_inference/CCommandParser.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include "CCommandParser.h"
+
+#include <core/CLogger.h>
+
+#include <rapidjson/error/en.h>
+#include <rapidjson/istreamwrapper.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/writer.h>
+
+#include <istream>
+
+namespace rapidjson {
+
+std::ostream& operator<<(std::ostream& os, const rapidjson::Document& doc) {
+    rapidjson::StringBuffer buffer;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+    doc.Accept(writer);
+    return os << buffer.GetString();
+}
+}
+
+namespace ml {
+namespace torch {
+
+const std::string CCommandParser::REQUEST_ID{"request_id"};
+const std::string CCommandParser::TOKENS{"tokens"};
+const std::string CCommandParser::VAR_ARG_PREFIX{"arg_"};
+
+CCommandParser::CCommandParser(std::istream& strmIn) : m_StrmIn(strmIn) {
+}
+
+bool CCommandParser::ioLoop(const TRequestHandlerFunc& requestHandler) {
+
+    rapidjson::IStreamWrapper isw{m_StrmIn};
+
+    while (true) {
+        rapidjson::Document doc;
+        rapidjson::ParseResult parseResult =
+            doc.ParseStream<rapidjson::kParseStopWhenDoneFlag>(isw);
+
+        if (static_cast<bool>(parseResult) == false) {
+            if (m_StrmIn.eof()) {
+                break;
+            }
+
+            LOG_ERROR(<< "Error parsing command from JSON: "
+                      << rapidjson::GetParseError_En(parseResult.Code())
+                      << ". At offset: " << parseResult.Offset());
+
+            return false;
+        }
+
+        if (validateJson(doc) == false) {
+            continue;
+        }
+
+        LOG_TRACE(<< "Inference command: " << doc);
+        jsonToRequest(doc);
+        if (requestHandler(m_Request) == false) {
+            LOG_ERROR(<< "Request handler forced exit");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool CCommandParser::validateJson(const rapidjson::Document& doc) const {
+    if (doc.HasMember(REQUEST_ID) == false) {
```

Comment:
Actually there is one security hole in the validation, which is that we need to confirm `doc[REQUEST_ID].IsString()`.  Without this additional check, sending an integer for this field instead would be a way to get a pointer of choice dereferenced.

Reply:
I added this and also the checks that the token arrays contain unsigned ints